### ISSUE TITLE
Unexpected error during compilation

### ISSUE THREAD

**Issue by GuillaumeLeclerc** [jashkenas/coffeescript#3926]
If compile the following code, the compiler crash unexpectedly :

```coffeescript
(toto : true) ->
    console.log "bug"

```

here is the partial call stack 

```
TypeError: Cannot read property 'length' of undefined
  at Param.exports.Param.Param.eachName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2123:30)
  at Param.exports.Param.Param.eachName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2126:16)
  at Code.exports.Code.Code.eachParamName (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2047:28)
  at Code.exports.Code.Code.compileNode (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:2001:12)
  at Code.exports.Base.Base.compileToFragments (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:83:21)
  at Block.exports.Block.Block.compileNode (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:387:28)
  at Block.exports.Block.Block.compileWithDeclarations (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:480:19)
  at Block.exports.Block.Block.compileRoot (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:452:24)
  at Block.exports.Block.Block.compileToFragments (/usr/local/lib/node_modules/coffee-script/lib/coffee-script/nodes.js:369:21)
  at /usr/local/lib/node_modules/coffee-script/lib/coffee-script/coffee-script.js:61:38
  at /usr/local/lib/node_modules/coffee-script/lib/coffee-script/coffee-script.js:34:19
```

If help is needed I might try to find some time and make a pull request.

Best regards and thank you for your help

Guillaume Leclerc

**Comment by christopherdumas** [comment]
What are you trying to do there? Because I don't think `arg : thing` is valid coffeescript.

**Comment by vendethiel** [comment]
I just think the compiler shouldn't crash, and report a parse error instead.

**Comment by akre54** [comment]
Nope, `argument: something` is a destructuring assignment ([example](http://coffeescript.org/#try:(a%3A%20b)%20-%3E)). @GuillaumeLeclerc, you're trying to assign `true = arg.toto`, which is an invalid lefthand side assignment.

**Comment by vendethiel** [comment]
Destructuring assignments need brackets

**Comment by akre54** [comment]
It is a bit confusing that `(a: b) ->` destructures, but `a: b = obj` does not.

**Comment by christopherdumas** [comment]
IMHO, I think that requiring brackets in this case would be a good thing.

**Comment by GuillaumeLeclerc** [comment]
I know this is not syntaxically correct, but I  think there should be an error instead an exception.



-----Message d'origine-----
De : "Adam Krebs" <notifications@github.com>
Envoyé : ‎03.‎04.‎2015 16:06
À : "jashkenas/coffeescript" <coffeescript@noreply.github.com>
Cc : "Guillaume Leclerc" <guillaume.leclerc.work@gmail.com>
Objet : Re: [coffeescript] Unexpected error during compilation (#3926)

It is a bit confusing that (a: b) -> destructures, but a: b = obj does not.

**Comment by christopherdumas** [comment]
@GuillaumeLeclerc yup!

**Comment by GuillaumeLeclerc** [comment]
If I understood well the error should be a name analysis because this is part of the grammar but it is an invalid assignement right ?

**Comment by adamdanielsdesign** [comment]
@GuillaumeLeclerc @vendethiel  @akre54 
```
#pretty sure
(toto : true) ->
    console.log "bug"
```
Should resolve to a compile time error, because it's an attempt to set a default to a argument.

**Comment by lydell** [comment]
How about this:

    $ coffee -bpe '(a:b)->'
    [stdin]:1:8: error: unexpected :
    (a:b->)
      ^

I could easily fix that after #4069 has been merged.

**Comment by vendethiel** [comment]
@lydell seems good!
